adapters/undertone/undertone.go

```diff
@@ -0,0 +1,197 @@
+package undertone
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/prebid/openrtb/v17/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"net/http"
+	"strconv"
+)
+
+const adapterId = 4
+const adapterVersion = "1.0.0"
+
+type adapter struct {
+	endpoint string
+}
+
+type BidRequestExt struct {
+	Prebid *openrtb_ext.ExtRequestPrebid `json:"prebid"`
+}
+
+type bidderParams struct {
+	Id      int    `json:"id"`
+	Version string `json:"version"`
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	imps, publisherId, errs := getImpsAndPublisherId(request)
+	if len(imps) == 0 {
+		return nil, errs
+	}
+
+	reqCopy := *request
+	reqCopy.Imp = imps
+
+	populateSiteApp(&reqCopy, publisherId, request.Site, request.App)
+	e := populateBidReqExt(&reqCopy)
+	if e != nil {
+		errs = append(errs, e)
+	}
+
+	requestJSON, err := json.Marshal(reqCopy)
+	if err != nil {
+		return nil, append(errs, err)
+	}
+
+	requestData := &adapters.RequestData{
+		Method: "POST",
+		Uri:    a.endpoint,
+		Body:   requestJSON,
+	}
+
+	return []*adapters.RequestData{requestData}, errs
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher. Run with request.debug = 1 for more info.",
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	impIdBidTypeMap := map[string]openrtb_ext.BidType{}
+	for _, imp := range request.Imp {
+		if imp.Banner != nil {
+			impIdBidTypeMap[imp.ID] = openrtb_ext.BidTypeBanner
+		} else if imp.Video != nil {
+			impIdBidTypeMap[imp.ID] = openrtb_ext.BidTypeVideo
+		}
+	}
```

Comment:
This is an anti-pattern. Also, this for loop code assumes that impression could either be Banner or Video and not both. But single impression can have both banner and video in the request. In that case, this code will just assumes that it is banner. 
Ideally, you should set response bid type based on bidder server response.

Reply:
hi @gargcreation1992 
In case our adserver selected an imp-id having both Video & Banner objects, the returned imp will be of Banner type